packages/mdc-textfield/_mixins.scss

```diff
@@ -62,3 +62,295 @@
     }
   }
 }
+
+// Private Mixins
+
+// Common
+
+@mixin mdc-text-field-invalid_ {
+  + .mdc-text-field-helper-text--validation-msg {
+    opacity: 1;
+    color: $mdc-text-field-error-on-light;
+  }
+
+  @include mdc-theme-dark(".mdc-text-field", true) {
+    + .mdc-text-field-helper-text--validation-msg {
+      color: $mdc-text-field-error-on-dark;
+    }
+  }
+
+  .mdc-text-field__label {
+    color: $mdc-text-field-error-on-light;
+  }
+
+  .mdc-text-field__input {
+    border-color: $mdc-text-field-error-on-light;
+  }
+
+  .mdc-text-field__bottom-line {
+    background-color: $mdc-text-field-error-on-light;
+  }
+
+  .mdc-text-field__idle-outline {
+    border-color: $mdc-text-field-error-on-light;
+  }
+
+  .mdc-text-field__input:hover ~ .mdc-text-field__idle-outline {
+    border-color: $mdc-text-field-error-on-light;
+  }
+
+  .mdc-text-field__outline .mdc-text-field__outline-path {
+    stroke: $mdc-text-field-error-on-light;
+  }
+}
+
+@mixin mdc-text-field-dense_ {
+  .mdc-text-field--dense {
+    margin-top: 12px;
+    margin-bottom: 4px;
+    font-size: .813rem;
+
+    .mdc-text-field__label--float-above {
+      // NOTE: This is an eyeball'd approximation of what's in the mocks.
+      transform: translateY(calc(-100% - 2px)) scale(.923, .923);
+    }
+  }
+}
+
+@mixin mdc-text-field-disabled_ {
+  .mdc-text-field--disabled {
+    pointer-events: none;
+
+    .mdc-text-field__input {
+      border-bottom: 1px dotted $mdc-text-field-disabled-border-on-light;
+
+      @include mdc-theme-dark(".mdc-text-field", true) {
+        border-bottom: 1px dotted $mdc-text-field-disabled-border-on-dark;
+      }
+    }
+
+    // stylelint-disable plugin/selector-bem-pattern
+    .mdc-text-field__input,
+    .mdc-text-field__label,
+    + .mdc-text-field-helper-text {
+      @include mdc-theme-prop(color, text-disabled-on-light);
+    }
+    // stylelint-enable plugin/selector-bem-pattern
+
+    .mdc-text-field__input,
+    .mdc-text-field__label {
+      @include mdc-theme-dark(".mdc-text-field") {
+        @include mdc-theme-prop(color, text-disabled-on-dark);
+      }
+    }
+
+    @include mdc-theme-dark(".mdc-text-field", true) {
+      // stylelint-disable plugin/selector-bem-pattern
+      + .mdc-text-field-helper-text {
+        @include mdc-theme-prop(color, text-disabled-on-dark);
+      }
+      // stylelint-enable plugin/selector-bem-pattern
+    }
+
+    .mdc-text-field__label {
+      bottom: 8px;
+      cursor: default;
+    }
+  }
+}
+
+@mixin mdc-text-field-focused_ {
+  .mdc-text-field--focused {
+    .mdc-text-field__input:required + .mdc-text-field__label::after {
+      color: $mdc-text-field-error-on-light;
+
+      @include mdc-theme-dark(".mdc-text-field", true) {
+        color: $mdc-text-field-error-on-dark;
+      }
+    }
+  }
+}
```

Comment:
As part of `mdc-text-field-focused` you should add this logic

```
.mdc-text-field--focused + .mdc-text-field-helper-text:not(.mdc-text-field-helper-text--validation-msg) {
  opacity: 1;
}
```

See original logic here:
https://github.com/material-components/material-components-web/blob/master/packages/mdc-textfield/mdc-text-field.scss#L727